Library/Homebrew/cask/lib/hbc/installer.rb

```diff
@@ -420,6 +437,19 @@ def zap
       purge_caskroom_path
     end
 
+    def backup_path
+      return nil if @cask.staged_path.nil?
+      if backed_up?
+        Pathname.new "#{@cask.staged_path}.upgrading"
+      else
+        @cask.staged_path
```

Comment:
I think this is really hard to follow. The only reason for this is `purge_versioned_files` using it. I'd suggest making two specialised versions of `purge_versioned_files`, one normal one, and one for the `upgrade?` case.

Reply:
I'll do it that way, and then factor out the whole metadata handling (to avoid code duplication).